src/Processors/QueryPlan/Optimizations/removeUnusedColumns.cpp

```diff
@@ -0,0 +1,196 @@
+}
+
+RemoveChildrenOutputResult removeChildrenOutputs(QueryPlan::Nodes & nodes, QueryPlan::Node & node)
+{
+    bool updated = false;
+    bool added_any_discarding_step = false;
+    const auto & parent_inputs = node.step->getInputHeaders();
+
+    for (auto child_id = 0U; child_id < node.children.size(); ++child_id)
+    {
+        auto & child_step = node.children[child_id]->step;
+        chassert(child_step->canRemoveUnusedColumns());
+
+        // Here we never want to remove inputs because the grandchildren might cannot remove outputs
+        const auto updated_anything
+            = child_step->removeUnusedColumns(getNameMultiSetFromNames(parent_inputs[child_id]->getNames()), false).updated_anything;
+
+        // TODO(antaljanosbenjamin): compare the header with name and types
+        // As removeUnusedColumns might leave additional columns in the output, we have get rid of those outputs by adding a new ExpressionStep
+        if (updated_anything)
+        {
+            const auto added_discarding_step = addDiscardingExpressionStepIfNeeded(nodes, node, child_id);
+            if (added_discarding_step)
+            {
+                chassert(node.children[child_id]->children.size() == 1 && (child_step) == node.children[child_id]->children[0]->step);
+
+#if defined(DEBUG_OR_SANITIZER_BUILD)
+                const auto & discarding_step = *node.children[child_id]->step;
+                assertBlocksHaveEqualStructure(
+                    *discarding_step.getInputHeaders()[0], *child_step->getOutputHeader(), "after adding discarding step");
+#endif
+                added_any_discarding_step = true;
+            }
+        }
+        /// Here child_step might have been replaced by a discarding step, so let's get the newest output headers
+#if defined(DEBUG_OR_SANITIZER_BUILD)
+        assertBlocksHaveEqualStructure(
+            *node.children[child_id]->step->getOutputHeader(), *parent_inputs[child_id], "after removing unused columns");
+#endif
+
+        if (updated_anything)
+            updated = true;
+    }
+
+    if (added_any_discarding_step)
+        return RemoveChildrenOutputResult::AddedDiscardingStep;
+
+    if (updated)
+        return RemoveChildrenOutputResult::Updated;
+
+    return RemoveChildrenOutputResult::NotUpdated;
+}
+
+size_t tryRemoveUnusedColumns(QueryPlan::Node * node, QueryPlan::Nodes & nodes, const Optimization::ExtraSettings &)
+{
+    auto logger = getLogger("removeUnusedColumns");
+    auto & parent = node->step;
+
+    auto max_updated_depth = 0U;
+
+    const auto & parent_inputs = parent->getInputHeaders();
+
+    for (auto child_id = 0U; child_id < node->children.size(); ++child_id)
+    {
+        auto current_update_depth = 0U;
+        auto * child_node = node->children[child_id];
+        auto & child_step = child_node->step;
+
+        if (!child_step->canRemoveUnusedColumns())
+            continue;
+
+        const auto can_remove_inputs = canAllChildrenCanRemoveOutputs(*child_node);
```

Comment:
It looks we better apply the optimization while walking bottom up though the plan, - this way we can avoid checking children.